### PR TITLE
Selective tests hash

### PR DESCRIPTION
**Pull request by fortmarek**
Resolves https://github.com/tuist/tuist/issues/1828

### Short description 📝

Let me explain quickly how it currently works:

I first hash the whole graph with `TestsGraphContentHasher`. Logic is *very similar* to `GraphContentHasher` with two differences 
- output type is `.none` because I just need to save the hash and there is nothing I'd put into the directory
- `isCacheable` logic is different because we don't to be limited to frameworks only

The hashes of `TestsGraphContentHasher` are saved to a temporary directory.

`TestsCacheGraphMapper` then goes through all the testable targets and determines if they have changed from the last successful run (by looking at `~/.tuist/Cache/TestsCache` and seeing if a hash of this target plus all of its dependencies is present there). I then delete from the schemes' test actions those targets that have been deemed cached. If no targets are left in `testAction`.

Run the tests with the "tree-shaken" graph.

If the tests are successfully completed, I move all the hashes from the temporary directory created early in the process and move them to `~/.tuist/Cache/TestsCache`. This ensures that the hashes are present there iff they have been a part of a successful test run - this part is also only done if all project schemes have been run (user has not specified any exact scheme) - otherwise we'd save hashes for test targets that have not actually been run.

In the original issue, it was proposed to do this optimization only if `--changed-modules` flag is passed. I don't think this is necessary and we can make this optimization by default. Alternatively, I'd be up to adding `-all-modules` that would ignore the cache but I'd do so only if there would be some demand for it.

I was testing this with `ios_app_with_tests` fixture - to test this locally, you can run `tuist test` twice and on the second run you should see that most of the tests have been skipped (I don't currently cache UI tests because their results are based on the destination device)

### Checklist ✅

- [x] The code architecture and patterns are consistent with the rest of the codebase.
- [ ] A developer other than the author has verified that the changes work as expected.
- [x] The changes have been tested following the [documented guidelines](https://tuist.io/docs/contribution/testing-strategy/).
- [x] The `CHANGELOG.md` has been updated to reflect the changes. In case of a breaking change, it's been flagged as such.
- [x] In case the PR introduces changes that affect users, the documentation has been updated.

### CODE REVIEW

**Review by kwridan — COMMENTED** [review]
Thanks @fortmarek 

I guess a gotcha that may need to be added to an FAQ for automation once this is released is about recommending that any custom build settings for paths defined in manifests be prefixed with `$(SRCROOT)/` (e.g. `SWIFT_OBJC_BRIDGING_HEADER`, I'm sure there may be others too which we may not be able to account for or automatically deal with).

File: Sources/TuistGenerator/Settings/DefaultSettingsProvider.swift
Comment (fortmarek) [reply Sources/TuistGenerator/Settings/DefaultSettingsProvider.swift:91]: sorry, we talked about this and I forgot to implement it

File: Sources/TuistGenerator/Generator/LinkGenerator.swift
Comment (fortmarek) [reply Sources/TuistGenerator/Generator/LinkGenerator.swift:234]: Is it possible to use `$(SRCROOT)` for input paths, too? I am not sure if that's the case. If so, I'll refactor it 🙂

**Comment by fortmarek** [comment]
> I guess a gotcha that may need to be added to an FAQ for automation once this is released is about recommending that any custom build settings for paths defined in manifests be prefixed with $(SRCROOT)/

Definitely agree - I had to fix our company project that I was testing this on where we used `PROJECT_DIR` as that is quite often used interchangeably with `SRCROOT`

File: Sources/TuistGenerator/Generator/LinkGenerator.swift
Comment (dharr419-ford) [reply Sources/TuistGenerator/Generator/LinkGenerator.swift:234]: It should be

File: Sources/TuistKit/Services/TestService.swift
Comment (dharr419-ford) [inline Sources/TuistKit/Services/TestService.swift:116]: Little confused as to what's going on here, as it almost reads like this is just an unused variable

**Review by dharr419-ford — COMMENTED** [review]
This is real nice

File: Sources/TuistKit/Services/TestService.swift
Comment (fortmarek) [reply Sources/TuistKit/Services/TestService.swift:116]: true, I'll rewrite it a little to make it less confusing 👍

File: Sources/TuistGenerator/Generator/LinkGenerator.swift
Comment (fortmarek) [reply Sources/TuistGenerator/Generator/LinkGenerator.swift]: fixed 👍

File: Sources/TuistGenerator/Settings/DefaultSettingsProvider.swift
Comment (fortmarek) [reply Sources/TuistGenerator/Settings/DefaultSettingsProvider.swift]: I have created `SourceRootPathProjectMapper`, so this is applied only for automation tasks

**Comment by fortmarek** [comment]
I believe this is ready for proper review, I have added tests, documentation, etc.

File: Sources/TuistAutomation/ProjectMappers/SourceRootPathProjectMapper.swift
Comment (pepibumur) [inline Sources/TuistAutomation/ProjectMappers/SourceRootPathProjectMapper.swift:18]: This doesn't prevent developers from overriding the value at the target level. In that case, the selective testing wouldn't work. I wonder if we should set that value at the target level instead and ignore whatever value is set. Since this mapper is only used for selective testing, I think it's fine. We can print a log telling developers that we are doing that.

File: Sources/TuistCache/ContentHashing/TestsGraphContentHasher.swift
Comment (pepibumur) [inline Sources/TuistCache/ContentHashing/TestsGraphContentHasher.swift:46]: I'd recommend changing the logic of the hashers to be cache agnostic. That'd mean removing this argument and also the internal checks that check whether the target is a framework or not. What you can do though to speed things up since you don't want to hash all the targets is passing a function that takes a target as an input and returns a boolean indicating whether the target should be hashed or not.

File: Sources/TuistCore/Cache/CacheOutputType.swift
Comment (pepibumur) [inline Sources/TuistCore/Cache/CacheOutputType.swift:21]: As I suggested in another comment, I would not introduce this.

File: Sources/TuistCache/ContentHashing/TestsGraphContentHasher.swift
Comment (fortmarek) [reply Sources/TuistCache/ContentHashing/TestsGraphContentHasher.swift:46]: > I'd recommend changing the logic of the hashers to be cache agnostic.

do you mean to strip `cacheOutputType` from `TargetContentHasher`? And for the `tuist cache` rehashing the output from `TargetContentHasher` with the `cacheOutputType` which is necessary to differentiate between `.xcframework` and `.framework`?

Or do you mean to merge `TestsGraphContentHasher` and `GraphContentHasher`? The logic for those is a little bit different and passing a simple function will not do. What I'd have to do is go through unit tests target and their dependencies, save them into array, and then in the passed function check if the target is in that array.

But I am not sure if that approach is very efficient + we actually would not reuse a lot of logic anyway, so I am not sure if such approach is worth it (I'd prefer just removing `cacheOutputType` from `TargetContentHasher`)

> also the internal checks that check whether the target is a framework or not.

do you mean checks in `GraphContentHasher`? Because there are no such checks for `TestsGraphContentHasher` and `TargetContentHasher`.

File: Sources/TuistAutomation/ProjectMappers/SourceRootPathProjectMapper.swift
Comment (fortmarek) [reply Sources/TuistAutomation/ProjectMappers/SourceRootPathProjectMapper.swift:18]: > This doesn't prevent developers from overriding the value at the target level.

I believe it's fine if users override this value. The original issue is that `SRCROOT` was equal to `PROJECT_DIR` but in this case it would stay unchanged. So if your "normal" project builds, so should the one for selective testing as the `SRCROOT` will be the same for both

**Comment by pepibumur** [comment]
One suggestion that I'd make @fortmarek is adding a markdown file under a `decision-records` directory where you explain at a high-level how this feature works. It'll be useful for someone in the future to gain context quicker without having to dive into all the components that are involved in making this happen. You can name the file `2021-02-07-selective-testing.md` and use the following structure:

```md
---
title: Selective testing
who: @fortmarek
---

## Context
What motivated us to implement this feature

## What
What's selective testing? Who would benefit from this feature? Why would you use selective testing in your workflows?

## How
Overview of how it's been implemented.

## Alternatives
Without selective testing, how can achieve the same results of speeding up test runs?
```
For now, we can keep them as markdown in the repository, but in the future we can even add a section to the website, allowing users to search across all our decision records cc @julianpanzer / @andreacipriani

File: Tests/TuistKitTests/Services/TestServiceTests.swift
Comment (pepibumur) [inline Tests/TuistKitTests/Services/TestServiceTests.swift:30]: You'll have to update the tearDown to set this one to nil. Otherwise, it'll remain allocated in memory until the completion of the entire test suite.

File: decision-records/2021-02-07-selective-testing.md
Comment (pepibumur) [inline decision-records/2021-02-07-selective-testing.md:8]: I'd remove the quotes from the sections.

File: Sources/TuistCache/ContentHashing/TestsGraphContentHasher.swift
Comment (pepibumur) [reply Sources/TuistCache/ContentHashing/TestsGraphContentHasher.swift:46]: What I mean is that I'd modify the hasher to have no knowledge on the cache. If you look at the implementation, internally we check if the target is cacheable in order to hash it. In order to make it reusable, I think we can remove that piece, and rather pass a function that allows filtering. That way we can control from what targets we'd like to hash. Like in your case, you are only interested in hashing the targets that are impacted by the selective testing. 
I imagine an API along the lines of:

```swift
func hash(graph: ValueGraph, take: (ValueGraphTarget) -> Bool) -> [ValueGraphTarget: String]
```

Notice the `take` attribute that allows the consumer of the API to choose the targets.

**Review by pepibumur — CHANGES_REQUESTED** [review]
Hey @fortmarek 
First of all, I really like the direction of the feature. It's awesome to see the solid mental model you've formed around Tuist and its architecture. I left some comments around some decisions you've made that I think require further discussion.

File: decision-records/2021-02-07-selective-testing.md
Comment (pepibumur) [inline decision-records/2021-02-07-selective-testing.md:22]: I'd not mention `$TUIST_CACHE_DIRECTORY` because it's an internal variable that allows us to control the directory for testing purposes.

Comment (fortmarek) [reply decision-records/2021-02-07-selective-testing.md:22]: `TUIST_CACHE_DIRECTORY` is actually now [documented](https://github.com/tuist/tuist/pull/2450#issuecomment-774531386) to allow users for custom location of the cache on CI

File: Sources/TuistKit/Commands/TestCommand.swift
Comment (fortmarek) [reply Sources/TuistKit/Commands/TestCommand.swift:51]: I'd agree but since there are changes between the project for testing and the one users actually use, it's quite possible there will be more issues - and without this there is no way they could debug this - it's also necessary for the acceptance tests. I can hide this with some env variable if we don't want it to be part of the command itself

Comment (fortmarek) [reply Sources/TuistKit/Commands/TestCommand.swift:51]: OK, sorry - I did not finish reading all the comments 😅 

[This](https://github.com/tuist/tuist/pull/2422#discussion_r571585238) should alleviate that need for users (we can print the path the project is generated to). I will still need to think about whether this will work for acceptance tests, too

File: Sources/TuistSupport/Utils/Environment.swift
Comment (fortmarek) [reply Sources/TuistSupport/Utils/Environment.swift:126]: we store the hashes of the targets that were part of a successful test run. I will think about better naming, I can't think of any right now

File: Sources/TuistKit/Services/TestService.swift
Comment (fortmarek) [reply Sources/TuistKit/Services/TestService.swift:94]: yes, the suggestion sounds reasonable to me 👍

File: Sources/TuistAutomation/ProjectMappers/SourceRootPathProjectMapper.swift
Comment (fortmarek) [reply Sources/TuistAutomation/ProjectMappers/SourceRootPathProjectMapper.swift:18]: @pepibumur does this make sense to you or are you still of the opinion that we should change this?

File: Sources/TuistCache/ContentHashing/TestsGraphContentHasher.swift
Comment (fortmarek) [reply Sources/TuistCache/ContentHashing/TestsGraphContentHasher.swift:46]: @pepibumur are you OK with creating `CacheGraphContentHasher` so the logic of filtering and hashing different output types could be reused? It will leverage `GraphContentHasher`, which will be cache agnostic.

I'm rather asking in advance because this change means a little more of moving around files and tests, so I don't want to waste that effort.

File: Sources/TuistKit/Services/TestService.swift
Comment (fortmarek) [reply Sources/TuistKit/Services/TestService.swift:94]: > I'd create a temporary directory

temporary as in it should be deleted after `tuist test` is finished? If so, this will not fix the issue of not being able to debug the test project. Instead of creating `automationPath`, we could introduce an env variable `TUIST_AUTOMATION_PATH` which would be needed for acceptance tests - and also with possible debugging in the future.

So, basically, I see two options:
- we generate test project to Tuist's global directory but without it being temporary
- we generate test project the same way we do now and allow configuration via env variable

If we generated test project to Tuist's global directory and deleted after command ended, it would have no advantage to the way we do it now (macOS temp directory)

Issue with generating project to Tuist's global directory without it being temporary is the fact that the directory might bloat in size quite quickly. Therefore, I'd be for hiding integrating automation path and using it only for acceptance tests / *our* debugging

What do you think, @pepibumur?

File: Sources/TuistSupport/Utils/Environment.swift
Comment (fortmarek) [reply Sources/TuistSupport/Utils/Environment.swift:126]: I don't think changing the name makes sense -> or change all the other directories located in cache which also have `Cache` in their name. But honestly, I think it's better as is because in code is used as `Environment.shared.testsCacheDirectory`. In that place you don't have to know that the `testsDirectory` is actually in `Cache` directory, so this way is imho more explicit and clearer.

File: Sources/TuistKit/Services/TestService.swift
Comment (fortmarek) [reply Sources/TuistKit/Services/TestService.swift:72]: does this bring any value over a more lightweight closure? I can make a `typealias` if typealias is what you are worried about. Otherwise, I don't see really what this brings but maybe I am missing something? And with a protocol I'll have to create an implementation class, mock - that's a lot of boilerplate for just a bunch of lines of code

**Comment by fortmarek** [comment]
Hey @pepibumur - I have answered some of your concerns and need some feedback before going forward. Thanks!

File: Sources/TuistKit/Services/TestService.swift
Comment (pepibumur) [reply Sources/TuistKit/Services/TestService.swift:94]: Here's what I'd suggest:
- Generate it in a temporary directory that gets deleted afterward.
- If you want to debug the generated project, you can do `tuist test --debug`, and that'll generate the project in a temporary directory and open it in Xcode for debugging. Once the Tuist process finishes, the project gets deleted.
- If we need to generate the project in a specific directory for the purpose of writing acceptance tests, then we can control that through an environment variable.

File: Sources/TuistSupport/Utils/Environment.swift
Comment (pepibumur) [reply Sources/TuistSupport/Utils/Environment.swift:126]: Fair enough

File: Sources/TuistKit/Services/TestService.swift
Comment (pepibumur) [reply Sources/TuistKit/Services/TestService.swift:72]: It's mainly being consistent with the patterns we've been using across the codebase. We achieve the same thing, but we are consistent, which at the end of the day is very appreciated when working on a large codebase. We can also provide mocks by conforming the protocol of that factory.

File: Sources/TuistCache/ContentHashing/TestsGraphContentHasher.swift
Comment (pepibumur) [reply Sources/TuistCache/ContentHashing/TestsGraphContentHasher.swift]: > @pepibumur are you OK with creating CacheGraphContentHasher so the logic of filtering and hashing different output types could be reused? It will leverage GraphContentHasher, which will be cache agnostic.

I'm not sure if I follow this one.

File: Sources/TuistAutomation/ProjectMappers/SourceRootPathProjectMapper.swift
Comment (pepibumur) [reply Sources/TuistAutomation/ProjectMappers/SourceRootPathProjectMapper.swift:18]: It does! Thanks for clarifying it.

File: Sources/TuistCache/ContentHashing/TestsGraphContentHasher.swift
Comment (pepibumur) [reply Sources/TuistCache/ContentHashing/TestsGraphContentHasher.swift]: I re-read it again and I understood what you mean. I think it's good to have that `CacheGraphContentHasher` for now.

**Review by pepibumur — APPROVED** [review]

File: Sources/TuistKit/Services/TestService.swift
Comment (fortmarek) [reply Sources/TuistKit/Services/TestService.swift:72]: that's a good point, I'll refactor it 👍

**Comment by fortmarek** [comment]
All comments have been addressed and the pipeline is green, will wait for another review

**Pull request merged**
